core/src/main/java/hudson/model/AutoCompletionCandidates.java

```diff
@@ -127,8 +128,10 @@ public void onItem(Item i) {
                  && i.hasPermission(Item.READ)
                     // and read permission required
                 ) {
-                    if (type.isInstance(i) && n.startsWith(value))
+                    if (type.isInstance(i) && n.startsWith(value) &&
+                            (!(container instanceof DirectlyModifiableTopLevelItemGroup) || !(i instanceof TopLevelItem) || ((DirectlyModifiableTopLevelItemGroup) container).canAdd((TopLevelItem) i))) {
```

Comment:
@jglick Could we use `DescriptorVisibilityFilter` to determine if it is a good candidate?